lib/cache.js

```diff
@@ -33,67 +36,72 @@ export function useDispatch() {
   return useReducer(cache, reducer)
 }
 
-export function useCache(resourceDescriptor) {
-  const {
-    serviceName,
-    queryId,
-    method,
-    id,
-    params,
-    realtime,
-    selectData,
-    transformResponse,
-    matcher,
-  } = resourceDescriptor
-
-  const dispatch = useDispatch()
-
-  const cachedData = useSelector(() => {
-    const query = getIn(cache(), ['queries', serviceName, queryId])
+const dataSelector = selector(
+  queryId => {
+    const { queries, entities, lookups } = cache()
+    const serviceName = getIn(lookups, ['serviceNamesByQueryId', queryId])
+    const query = getIn(queries, [serviceName, queryId])
     if (query) {
-      const { data } = query
-      const entities = query.entities || getIn(cache(), ['entities', serviceName])
-      return selectData(data.map(id => entities[id]))
+      const items = query.entities || entities[serviceName]
+      return query.selectData(query.data.map(id => items[id]))
     } else {
       return null
     }
-  }, [serviceName, queryId, selectData])
+  },
+  { label: 'figbird:data', persist: false },
+)
 
-  const cachedResult = useSelector(() => {
-    const query = getIn(cache(), ['queries', serviceName, queryId])
+const metaSelector = selector(
+  queryId => {
+    const { queries, lookups } = cache()
+    const serviceName = getIn(lookups, ['serviceNamesByQueryId', queryId])
+    const query = getIn(queries, [serviceName, queryId])
     if (query) {
-      const { meta } = query
-      return { ...meta, data: cachedData }
+      return query.meta
     } else {
-      return { data: null }
+      return null
     }
-  }, [serviceName, queryId, cachedData])
+  },
+  { label: 'figbird:meta', persist: false },
+)
 
+export const querySelector = selector(
+  queryId => {
+    const data = dataSelector(queryId)
+    const meta = metaSelector(queryId)
```

Comment:
Both these selectors are only used here, and have some duplicate lookups in them, so maybe they could just be inlined here?

Reply:
Wouldn't work the same, because dataSelector computes a fresh array each time by resolving query IDs to actual objects. But if no items actually changed the dataSelector will return a memoised value which will avoid re-rendering downstream. Does that make sense?

It's a bit how you'd use multiple nested [reselect](https://github.com/reduxjs/reselect) selectors to avoid busting memoisation.

It's a bit hidden / unclear.. but it's all about avoiding those re-renders.. Or in fact, it might be this could cause an infinite re-render without it, because every time you render you'd get a fresh data array, which could trigger an effect (I observed this in our app where there's a selector selecting some people, and then has a useEffect->setState if people have changed).

Reply:
Ah yep, because internally it checks if the array is deep equal 👍